lib/app/modules/bulk_data/bulk_data_discovery_sequence.rb

```diff
@@ -0,0 +1,173 @@
+
+      def oauth2_metadata_from_conformance
+        options = {
+          token_url: nil
+        }
+        begin
+          @conformance.rest.each do |rest|
+            options.merge! @client.get_oauth2_metadata_from_service_definition(rest)
+          end
+        rescue StandardError => e
+          FHIR.logger.error "Failed to locate SMART-on-FHIR OAuth2 Security Extensions: #{e.message}"
+        end
+
+        options.delete_if { |_k, v| v.nil? }
+        options
+      end
+
+      test :read_well_known_endpoint do
+        metadata do
+          id '01'
+          name 'Retrieve Configuration from well-known endpoint'
+          link 'http://www.hl7.org/fhir/smart-app-launch/conformance/#using-well-known'
+          description %(
+            The authorization endpoints accepted by a FHIR resource server can
+            be exposed as a Well-Known Uniform Resource Identifier
+          )
+          optional
+        end
+
+        well_known_configuration_url = @instance.url.chomp('/') + '/.well-known/smart-configuration'
+        well_known_configuration_response = LoggedRestClient.get(well_known_configuration_url)
+        assert_response_ok(well_known_configuration_response)
+        assert_response_content_type(well_known_configuration_response, 'application/json')
+
+        @well_known_configuration = assert_valid_json(well_known_configuration_response.body)
+
+        @well_known_token_url = @well_known_configuration['token_endpoint']
+        @well_known_register_url = @well_known_configuration['registration_endpoint']
+        @well_known_token_auth_methods = @well_known_configuration['token_endpoint_auth_methods_supported']
+        @well_known_token_auth_signings = @well_known_configuration['token_endpoint_auth_signing_alg_values_supported']
+        @well_known_scopes = @well_known_configuration['scopes_supported']
+
+        @instance.update(
+          oauth_token_endpoint: @well_known_token_url,
+          oauth_register_endpoint: @well_known_register_url
+        )
+      end
+
+      test :validate_well_known_configuration do
+        metadata do
+          id '02'
+          name 'Well-Known Configuration contains required fields'
+          link 'https://build.fhir.org/ig/HL7/bulk-data/authorization/index.html#advertising-server-conformance-with-smart-backend-services'
+          description %(
+            The JSON from .well-known/smart-configuration contains the following
+            required fields: #{WELL_KNOWN_FIELDS.map { |field| "`#{field}`" }.join(', ')}
+          )
+          optional
+        end
+
+        skip 'Server does NOT provide .well-known endpoint' unless @well_known_configuration.present?
+
+        missing_fields = WELL_KNOWN_FIELDS - @well_known_configuration.keys
+        assert missing_fields.empty?, "The following required fields are missing: #{missing_fields.join(', ')}"
+      end
+
+      test :read_conformance_oauth_endpoins do
```

Comment:
All the servers tested are using the CapabilityStatement not .well-known endpoint.

Reply:
So is this an oversight in the IG or are the servers not conforming to the IG?

Reply:
I think this is an IG issue. SMART-On-FHIR has a whole section talking about endpoints in CapabilityStatement and most bulk-data servers are also multi-purposed as SOF servers. I think BulkData IG SHOULD also mention that. I have sent a message to Josh about adding Conformance into BulkData IG.

Reply:
Also BulkData IG says server MAY advertise using well-known endpoints.

Reply:
That MAY just confuses me since they don't say anything about the CapabilityStatement. Does that mean they don't have to advertise at all? The SMART app launch IG says servers SHALL use both.